ocaml/typing/mode.ml

```diff
@@ -2,1242 +2,1750 @@
-    let uniqueness = Uniqueness.constrain_lower t.uniqueness in
-    let linearity = Linearity.constrain_lower t.linearity in
-    { locality; uniqueness; linearity }
+    let meet (l0, l1, l2) (r0, r1, r2) =
+      ( Locality.Const.meet l0 r0,
+        Linearity.Const.meet l1 r1,
+        Uniqueness.Const.meet l2 r2 )
 
-  let newvar () =
-    let locality = Regionality.newvar () in
-    let uniqueness = Uniqueness.newvar () in
-    let linearity = Linearity.newvar () in
-    { locality; uniqueness; linearity }
-
-  let newvar_below { locality; uniqueness; linearity } =
-    let locality, changed1 = Regionality.newvar_below locality in
-    let uniqueness, changed2 = Uniqueness.newvar_below uniqueness in
-    let linearity, changed3 = Linearity.newvar_below linearity in
-    { locality; uniqueness; linearity }, changed1 || changed2 || changed3
-
-  let newvar_above { locality; uniqueness; linearity } =
-    let locality, changed1 = Regionality.newvar_above locality in
-    let uniqueness, changed2 = Uniqueness.newvar_above uniqueness in
-    let linearity, changed3 = Linearity.newvar_above linearity in
-    { locality; uniqueness; linearity }, changed1 || changed2 || changed3
-
-  let check_const t =
-    let locality = Regionality.check_const t.locality in
-    let uniqueness = Uniqueness.check_const t.uniqueness in
-    let linearity = Linearity.check_const t.linearity in
-    { locality; uniqueness; linearity }
-
-  let print' ?(verbose = true) ppf t =
-    Format.fprintf ppf "%a, %a, %a"
-      (Regionality.print' ~verbose ~label:"locality")
-      t.locality
-      (Uniqueness.print' ~verbose ~label:"uniqueness")
-      t.uniqueness
-      (Linearity.print' ~verbose ~label:"linearity")
-      t.linearity
-
-  let print ppf t = print' ~verbose:true ppf t
+    let join (l0, l1, l2) (r0, r1, r2) =
+      ( Locality.Const.join l0 r0,
+        Linearity.Const.join l1 r1,
+        Uniqueness.Const.join l2 r2 )
+
+    (** constrain uncurried function ret_mode from arg_mode *)
+    let close_over (locality, linearity, uniqueness) =
+      let locality' = locality in
+      (* uniqueness of the returned function is not constrained *)
+      let uniqueness' = Uniqueness.Const.min in
+      let linearity' =
+        Linearity.Const.join linearity
+          (* In addition, unique argument make the returning function once.
+             In other words, if argument <= unique, returning function >= once.
+             That is, returning function >= (dual of argument) *)
+          (Const.unique_to_linear uniqueness)
+      in
+      locality', linearity', uniqueness'
+
+    (** constrain uncurried function ret_mode from the mode of the whole function *)
+    let partial_apply (locality, linearity, _) =
+      let locality' = locality in
+      let uniqueness' = Uniqueness.Const.min in
+      let linearity' = linearity in
+      locality', linearity', uniqueness'
+  end
+
+  let close_over comonadic monadic =
```

Comment:
Instead of taking these two arguments separately, just take a `t`.

Reply:
The two are separated, because they have opposite `allowed/disallowd` status, which is caused by the `unique_to_linear` morphism flipping the status of a mode, which is caused by monadic dual to comonadic.

Reply:
I've done this one.